openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java

```diff
@@ -154,62 +163,174 @@ public void handlePayloadAnnotation(RequiredParameterNode payloadNode, Map<Strin
+                    OptionalTypeDescriptorNode optionalTypeDescriptorNode = (OptionalTypeDescriptorNode) payloadNode;
+                    handlePayloadType((TypeDescriptorNode) optionalTypeDescriptorNode.typeDescriptor(),
+                            schema, bodyParameter);
+                    break;
+                case UNION_TYPE_DESC:
+                    UnionTypeDescriptorNode unionTypeDescriptorNode = (UnionTypeDescriptorNode) payloadNode;
+                    TypeDescriptorNode leftTypeDesc = unionTypeDescriptorNode.leftTypeDesc();
+                    TypeDescriptorNode rightTypeDesc = unionTypeDescriptorNode.rightTypeDesc();
+                    handlePayloadType(leftTypeDesc, schema, bodyParameter);
+                    handlePayloadType(rightTypeDesc, schema, bodyParameter);
+                    break;
+                case SIMPLE_NAME_REFERENCE:
+                    SimpleNameReferenceNode record = (SimpleNameReferenceNode) payloadNode;
+                    TypeSymbol typeSymbol = getReferenceTypeSymbol(semanticModel.symbol(record));
+                    String recordName = record.name().toString().trim();
+                    handleReferencePayload(typeSymbol, mediaTypeString, recordName, schema, bodyParameter);
+                    break;
+                case QUALIFIED_NAME_REFERENCE:
+                    QualifiedNameReferenceNode separateRecord = (QualifiedNameReferenceNode) payloadNode;
+                    typeSymbol = getReferenceTypeSymbol(semanticModel.symbol(separateRecord));
+                    recordName = ((QualifiedNameReferenceNode) payloadNode).identifier().text();
+                    TypeDescKind typeDesc = ((TypeReferenceTypeSymbol) typeSymbol).typeDescriptor().typeKind();
+                    String mimeType = getMediaTypeForTypeDecKind(typeDesc);
+
+                    handleReferencePayload(typeSymbol, mimeType, recordName, schema, bodyParameter);
+                    break;
+                case ARRAY_TYPE_DESC:
+                    ArrayTypeDescriptorNode arrayNode = (ArrayTypeDescriptorNode) payloadNode;
+                    mediaTypeString = getMediaTypeForSyntaxKind(arrayNode.memberTypeDesc());
+                    handleArrayTypePayload(schema, arrayNode, mediaTypeString, bodyParameter);
+                    break;
+                default:
+                    //Warning message for unsupported request payload type in Ballerina resource.
+                    IncompatibleResourceDiagnostic error = new IncompatibleResourceDiagnostic(
+                            DiagnosticMessages.OAS_CONVERTOR_117, payloadNode.location(),
+                            payloadNode.toSourceCode().trim());
+                    diagnostics.add(error);
+                    break;
+            }
+        } else {
+            //Warning message for unsupported request payload type in Ballerina resource.
+            IncompatibleResourceDiagnostic error = new IncompatibleResourceDiagnostic(
```

Comment:
Rather than using the `if-else` block i think w could check whether the condition failure and do an early return from there after adding the necessary warnings to the diagnostic-context.